### DISCUSSION TITLE
[CI] Mobile applications artifacts management

### DISCUSSION THREAD

**Discussion by frengooo** [mia-platform/community#154] [Q&A] [answered]
Hi guys,
we usually build and push on nexus containers of microservices (FE/BE).

How can we handle mobile applications artifacts instead?
E.g. in a CI pipeline can we push the Android `apk` (or the IoS equivalent) on a registry like Nexus?
Otherwise, how we handle the mobile artifacts?

Thanks.

**Comment by javferrod** [comment, marked as answer]
Hi!

There are different phases involved when deploying a mobile application:

- **Build**: You need an environment that allows you to build your application. Android can run wherever you want. For iOS, this usually involves running your CI in a macOS instance. You can set up a Mac runner in your CI provider or use external platforms such as [Bitrise](https://bitrise.io).
- **Testing**: If you want e2e tests with [Espresso](https://developer.android.com/training/testing/espresso), XCUITest or [Appium](https://appium.io/) you are going to need emulators or real devices. There are plenty of providers like [AWS Device Farm](https://aws.amazon.com/device-farm/), [Firebase](https://firebase.google.com/products/test-lab) or [Sauce Labs](https://saucelabs.com/platform/mobile-testing). You can always set up something on your own but it's a complex process.
- **Storage**: You can keep your binaries wherever you want: S3 bucket, nexus, CI artifacts... Usually, the binaries are stored using the platforms and tools specific to each platform. For Android, this would be [Firebase](https://firebase.google.com/products/app-distribution) or the Play Store and for iOS [Apple TestFlight](https://testflight.apple.com/).
- **Beta distribution:** Usually you want to create some kind of beta version to test your applications on real devices, this can be archived using Firebase or the Play Store for Android and Apple TestFlight and iOS App Store for iOS.
- **Public Distribution**: This phase requires using Google Play Store and iOS App Store. 

There are plenty of tools that help with the automatization of these steps. The most known open source option is [Fastlane](https://fastlane.tools/). [Bitrise](https://bitrise.io) it's also a good option If you are willing to pay.

Hope it helps! Regards.
